---
title: "Somatic mutations modulate ceRNA drivers of tumorigenesis"
author: Jing He
date: May, 09, 2014
output: pdf_document
---

```{r,echo=FALSE,message=FALSE,error=FALSE,warning=FALSE,prompt=FALSE}
# install.packages("devtools")
# devtools::install_github("rstudio/rmarkdown")
# library(devtools)
# install_github("knitcitations", "cboettig")
library(knitr)
library(markdown)
library(knitcitations)
library(bibtex)
write.bib(c('bibtex', 'knitr', 'knitcitations'), file="genReport.bib")
biblio <- read.bib("genReport.bib")
setwd("/Volumes/ifs/home/c2b2/ac_lab/jh3283/scripts/projFocus/ceRNA/report/")
setwd("/Volumes/ifs/scratch/c2b2/ac_lab/jh3283/projFocus/ceRNA/report/")
render("genReport.Rmd", "pdf_document")

```
Somatic mutations modulate ceRNA drivers of tumorigenesis
=============================================
**Jing He;Hua-sheng Chiu;Pavel Sumazin; Andrea Califano**

Abstract:

Pan-cancer studies have shown that competitive endogenous RNA (ceRNA) networks can cooperate with chromosome instability and abnormal DNA methylation in tumors to dysregulate tumor suppressors and oncogenes. However, ceRNA cooperative association with mutations in cancer has not been studied. Integrating data from TCGA and ENCODE, we show that the cooperation between ceRNA interactions and mutations of unknown function contribute to the dysregulation of cancer genes. 

We integrated ceRNA networks and mutations in an attempt to mechanistically recover missing genomic variability of cancer genes in TCGA breast cancer biopsies. Genes have missing genomic variability in a tumor dataset when their dysregulation cannot be explained through profiling of their DNA locus. Using a group lasso regression model we showed that ceRNA drivers cooperating with somatic mutations, CNV, and methylation, could account for a large fraction of the missing genomic variability of cancer genes in breast cancer.  Moreover, using a greedy-forward optimization algorithm, we identified ceRNA driver mutations that could potentially drive tumorigenesis through the ceRNA mechanism. Furthermore, we showed that driver ceRNA mutations are enriched in known and predicted binding sites of transcription factors and microRNAs. 

In summary, our results suggest that somatic mutations, often of unknown function, cooperate with ceRNA regulators to alter the expression of cancer genes in breast cancer tumors.


1. Background and introduction
-------------------------------
Pan-cancer studies has shown that mutations in cancer are frequent genetic variation which would be oncogeneic. Previous work has shown that ceRNA network can integrate genetic and epigenetic alterations, such as Copy Number Variabtion(CNV), methylations,  of cognate ceRNA regulators to dysregulate established oncogenes and tumor suppressors, accounting for a large fraction of the missing genomic variability in tumors. However, ceRNA coooperative association with mutations in cancer has not been studied.  

On the other hand, accumulating evidence showed that somatic mutations are frequent genomic abberrants, and play an important role in cancer ongenesis. Recurrent somatic mutations in genes account for specific subtypes of cancer. Several studies have shown that mutations in coding regions, especially non-synonmous ones, might cause bad prognosis of cancers. But the impact or function for mutations with low recurrence, or in non-coding regions are still not clear. 

Genes have missing genomic variability in a tumor dataset when their dysregulation cannot be explained through profiling of their DNA locus. 
In this study, we investigate the potential function of mutations in regulatory region, including transcription factor(TF) binding sites and microRNA binding sites, explored their regulatory ability throught ceRNA network. especially focuse on those in regulatory regions. we find taht cancer genes and tumor suppressors missing genomic variablity can be explained by regulatory mutations of it's cognate ceRNA dirver. 


This appoach is also based on hierarchical regression model that evaluates candidate group/aberrants, or ‘modulators’. The association coefficient measures how well a modulator (or combination of modulators) contribute to the behavior of ceRNA effection through gene expression across tumor samples. 


2. Data
------------------------------

TCGA data is publicly avaible, relatively has high quality, and most importly, have reasonable sufficient data of different level. All data are downloaded from TCGA, an access using lab key. 

- rnaSeq: version 1, level 3 raw counts was used in edgeR to get differential expressed genes(DEG), total sample size 778, normal sample 100
  
- somaticMutation:  level 2 maf file from tcga, total sample size 1001,

- CNV_snparray: level 2, total sample size 1049

- methylation: level 3 data from TCGA, both tumor and normal are used to get differential methylation genes. ( sample size ,tumor ,normal )

ENCODE data
3. Methods
---------------------------------

###  3.1 identify genomic intact dysregulated cancer genes(Gint) and samples
- Hypothesis 1, CNV, methylation, and somatic mutation can cause expression dysregulation of cancer genes. < experimental validated >
- Hypothesis 2. For those cancer genes which are dysregulated but did not show somatic mutations, CNV, methylations, we suspect that ceRNA might be accounting for their dysregulation < previous study computational validated>
  Are the any other type of regulations which can play this role? <epigenetic level regulatory, other forms of genetic abberrants which we haven't included into this study so far> 
  <But, for cancer patients, somatic mutations are more frequent, while other forms of genetic factors such as chromosome reorganization, 
- Hypothesis 3. ceRNA effect was affected by multiple factors: 1> ceRNA abundance, 2> microRNA abundance. 3> targe abundance . Genetic abberants of ceRNA can lead to ceRNA abundance change, which pertublent the ceRNA regulation on target cancer gene.

0) data pre-processing
- Differential expressed cancer genes
   - voom normalization for all normal sample
   - t.test for the different expression, 712 in 502 out
   - FC and bonferroni correction p-value: vocano plot 
- cancer genes' ceRNET regulators expression
  - 3793 regulators for 503 UCGgenes
  - EdgeR: using rowcounts as expression
- UCGR genomic aberrants :
  - SNP: get throught large sample size 

1). cancer genes got from ceRNA project.(n= 776) in UC-ceRNET
    
2). cnv sample elimination : 
    - renames all cnv files using barcode; get unique gene start, end site file; list all available cnv file(sample have cnv); segment mean is log2 ratio of copy number (0.8, 1.2)     copy number(1.74,2.3); for genes have multiple cnvs cover it, we choose the one have higher cnv value(either del or amp) [documentation](https://tcga-data.nci.nih.gov/tcgafiles/ftp_auth/distro_ftpusers/anonymous/tumor/read/cgcc/broad.mit.edu/genome_wide_snp_6/snp/broad.mit.edu_READ.Genome_Wide_SNP_6.mage-tab.1.2003.0/DESCRIPTION.txt)
    - some figures show number of samples for each gene


3). methylation sample elimination: 
    - tumor sample(n=1051), normal sample(n=115); for genes with multiple probes, we only use the probe that is unique to this gene. sva package in bioconductor was used to normalize out batch effect. 7 normal samples which show  as outlier were removed(n = 89). zscore was calculated between normalized tumore and normal samples, and genes with a z-score > 2 was considered as differential methylated 
    - R package(sva) was used to do batch normalization, all NA data was filled with mean value of this gene [sva]
    - all beta value was transformed to M value for differential methylation analysis
    
4). mutation sample 
    tcga somatic mutation level2 data
5). dysregulated genomic intact cancer genes and samples
  
###  3.2 identify candidate driver ceRNA regulator for each Gint
   
1). group lasso
    We are interested to find the explantory factors accounting for the response variables, where each explantory factor may be represented by a group a derived input variable. A classical lasso which was designed to select individual input variables, could not select multiple variables if they are co-ordinated. Instead, group lasso which allows group selection. 
    
    i). clustering 
    r package adegenet was used to do clustering. The clustering procedure consists in running successive K-means with an increasing number of clusters (k), after transforming data using a principal component analysis (PCA). For each model, Alkaike Information Criteria(AIC) is computed, the minimum AIC model was choosen as the model for optimal k.
    
    ii). regression 
      After assign a group to each of the input variables, a group lasso was using the Voom normalized expression data for both target and its ceRNA regulators. Both l1 group lasso penalty(lambda) and l2 ridge penalty(alpha) was applied. 100 lambda value was tried, and best model was selected according to AIC.  10 alpha values from 0 to 1 was tried, and the one which minimum regression rss(residual sum of square) was selected as the best one. A maximun of 5000 iteration was use, with an tolenrance of 0.05 (eps=0.05,max.iter=5000) for each alpha value iteration.
      
      A group lasso allows us to do a group level seletcion, which allows selection of collinear variables which are assigned to one group compared to classical lasso model.  
      
    iii). premutation
      After running lasso model on discovered groups, a residual permuatation precedure was used to compute significance level of this model. This process is to ensure the residual nomality and error indenpendence of the previous linear regression mode. A pvalue of 0.05 was used to discover any significant regression aftering 1000 permutation. < the coefficient determinant of R2 is significant >
    iv). test to check if selected mutation are significantly representative. (fisher exact test) (loose promoter to 2 kb)
      1. compare to results from lasso: (lasso number of ceRNA driver v.s greedy number of ceRNA driver)
      2. compare to 
4). filtering and get candidate ceRNA driver
    based on a permuation of 1000, a permutated r2 p-value < 0.01 indicate a significant ceRNA drivers regulation of their' cognate target genes. For thos target gene, all ceRNA regulators which have a regression coefficient > 0 and permuation pvalue < 0.01  were selected as ceRNA drivers.
  
  
###  3.3 identify driver mutations for ceRNA driver 
1). get all somatic mutation from TCGA level2, including 7 kinds of mutations

2). collapsing mutations by each gene as the annotation

3). identify mutated ceRNA driver mutations 

4). identify dirver mutation sets for each cancer target gene
  - Biological Question:
    - identify a subset of samples in which ceRNA drivers total expression correlated with target expression
    - compute the differential expresion of this set of samples versus non-mutated samples
  - formulaize question in Mathmatical word:
  
  - using greedy optmizaion of correlation to select samples and ceRNA drivers
  
  - fisher exact test to test if driver mutation/genes are significantly enriched in regulatory region(promoter region/3p UTR region)
        -- do this for whole set of genes
        -- do this only for cancer genes(starting from 984)

###  3.4 functional filtering of ceRNA driver mutations
    - binding sites mutations 
            TF binding sites mutations
            miRNA binding sites information 
    - ENCODE data 
            --peak and hotspot data are downloaded from ENCODE website. for availabe cell line of breast cancer
            treated v.s none, ctrl v.s none 
            
    - expression/activity information
      - impact of the selected genes and subsets. < choose the right control group>
      - single mutation in one ceRNA dirver not going to help
    - Dnase strong signal region
    - chromatin open region

###  3.5 hub analysis 
     idea: network analysis, find the nodes of diver/target with hight degrees/centrality 
    - targets share common dirvers
    - common drivers share targets


###  3.6 validation of functional ceRNA driver mutations (TBD)

4. Results
-------------------------------------------------------
1). Identify genomic intact samples and cancer genes(Gint)
    - 593 cancer genes were identified as dysregulated genomic intact genes for downstream analysis
    - big venn diagram < figure 1> 
    
2). Identify candidate driver regulators for each Gint
    - MBD2; MAX;has no effective regulator; DDX3X is still running..
    - 343 cancer genes are identified as having at least one driver regulators, 2749 unique candidate driver regulators
    - summerizing file(/ifs/data/c2b2/ac_lab/jh3283/projFocus/result/03102014/candiReg/run-Apr-1-2014/kegRegs_Apr-16-2014.summary)
    - example of one gene ( heatmap, )
    
3) Identify functional ceRNA driver somatic mutation

    - Step 1
      TCGA level 2 BRCA somatic mutations(both curated and validated) 95371 mutations(locus) in at least 1 patients(data in /ifs/data/c2b2/ac_lab/jh3283/projFocus/result/03102014/tcgal2som )      
      18,828 in promoter Regions; 
      7,417 3' utr Regions
    - step 2
      9,829 regulatory mutated genes
    - step 3 (if run functional regulatory mutaiton at first)
      72 mutated driver regulators
    - step 4
    
4)  mutations in ceRNA TF binding sites and microRNA binding sites enriched for cancer genes
    
      
5. Discussion
-------------------------------------------------------
    -  
  
  
8. Acknowledgement 
-------------------------------------------------------



9. Reference
-------------------------------------------------------

[sva]: http://bioinformatics.oxfordjournals.org/content/early/2012/01/17/bioinformatics.bts034
[ck]: http://en.wikipedia.org/wiki/Cohen's_kappa
[kw]: http://en.wikipedia.org/wiki/Kruskal%E2%80%93Wallis_one-way_analysis_of_variance
[hwe]:http://en.wikipedia.org/wiki/HWE
[bs]: http://www.broadinstitute.org/science/programs/medical-and-population-genetics/birdsuite/birdsuite-analysis#birdsuite_snps
[lifeOver]: http://genome.cshlp.org/cgi/pmidlookup?view=long&pmid=12045153
[cluster]:Jombart, Thibaut, Sébastien Devillard, and François Balloux. "Discriminant analysis of principal components: a new method for the analysis of genetically structured populations." BMC genetics 11.1 (2010): 94.
[residual permutation]: Potter, Douglas M. "A permutation test for inference in logistic regression with small‐and moderate‐sized data sets." Statistics in medicine 24.5 (2005): 693-708.