\documentclass[a4paper,11pt]{article}
\usepackage{listings}
\usepackage{indentfirst}
\usepackage{amsmath}
\usepackage{algorithm}
\usepackage{algorithmic}
\usepackage{graphicx}
\usepackage{setspace}
\usepackage{alltt}
\usepackage{graphicx}
\usepackage{float}
\usepackage{caption}
\usepackage{authblk}
\usepackage{tabularx} 
\usepackage{footnote}
\usepackage{Sweave}
\usepackage{hyperref}
\graphicspath{ {/Users/jh3283/Dropbox/} } 
\hypersetup{ colorlinks, linkcolor=blue}
 
% \doublespacing
\singlespacing
\title{Explain ceRNA effect from genomic variations}
\author{Jing He\,$^{1\,2*}$,Hus-Sheng}
\affil[1]{Department of Biomedical Informatics, Columbia Medical Center}
\affil[2]{Center for Computational Biology and Bioinformatics}

\renewcommand\Authands{ and }

\begin{document}
\input{draft-concordance}
\maketitle
% \begin{abstract}
% \end{abstract}

\section{Background and Questions}

Based on Hua-sheng's defense topic which excluded mutation, cnv for ceRNA effect
Question 1: SNPs/CNVs/Methylations/Somatic mutations regulate gene expressions, the expression of genes have effect on ceRNA effect, but how much does each of them contribute to the final ceRNA effect?

Question 2: 

Goal:
\begin{itemize}
	\item Find snps that are associated with DEGs, and explian it's effect using network approach
	\item Find indels that are associated with DEGs, and briding with ceRNA network
\end{itemize}

\section{Methods}

1.SNPs association with Gene Expression/indel associating with GeneExpression
\begin{itemize}
	\item SNP filtering process 1(total 906602):
		- SNP quality: out 78043
		- SNP MAF: out 98
		
	\item association study(total 828461) :
		- Cohen's Kappa  \url{http://en.wikipedia.org/wiki/Cohen's_kappa}
		- KW test( threshold to be confirmed)

	feature selection methodsï¼šgroup lasso
	hierachary structure:
2. TSS 
3. 3' UTR

\section{Data}
1. TCGA level2 SNP array
2. TCGA level3 CNV (SNP array)
3. TCGA level3 RNAseq
All BRCA samples who has the three types of data were selected (total 731 samples, finalized on )

Workflow
\subsection{} 


\subsection{}

\section{Validation}
UK10K
1000 Genome
Encode data 3C, 5C, H-G

\section{Conclusion}

\section*{Acknowledgement}

\bibliographystyle{plain}
\bibliography{draft} 

\section{Figures}

	% Figure:
		% 1. number of somatic mutations:
		% 	% show the different number of SMs(WGS, coding)
		% 2. Mutated Gene frequency and heatmap in all patients(show the diversity of mutated genes)
		% 3. Heterogeneity: number of colones and heatmap for mixture coefficience
		% 4. Shannon Index and Survivial analysis
		% 5. Network roubustness
		% 6. Table, 121 gene rank in different patients
		% 7. Enrichment plot for mapped genes &  Gene Ontoloy for unmapped Genes and mapped genes
		% 8. validation part: Venn diagramm showing the overlapped genes


\end{document}
